ManageIQ/manageiq-performance
# [WIP] Add profile sub command to miqperf

This command allows for a various benchmarking/profiling (memory, time, stackprof, etc.) to be introduced into an existing ruby script, without having to hardcode it into the script/codebase itself.

This assumes there is single entry point into the script, and creates a temp script to run that dynamically adds the necessary profiling utilities (with the minimum requirements to do so), and runs that in a spawned process.

Example
-------

To run a script and just see how much memory it uses, you can run the following:

```console
$ bundle exec ruby miqperf profile -m lib/foo.rb

TOTAL_MEMORY_USED: 170.95703125MiB
```

To run the same script, but see how long it took, use `-t`:

```console
$ bundle exec ruby miqperf profile -t lib/foo.rb

Timings
-------
real    0m6.781s
user    0m6.070s
sys     0m0.590s
cuser   0m0.000s
csys    0m0.000s
```

Or include both of the outputs together:

```console
$ bundle exec ruby miqperf profile -mt lib/foo.rb

TOTAL_MEMORY_USED: 170.95703125MiB

Timings
-------
real    0m6.781s
user    0m6.070s
sys     0m0.590s
cuser   0m0.000s
csys    0m0.000s
```

The output can also be set using the `-o`.  Use the `-h` option with this branch to see all of the available options


Links
-----
* Borrows portions of this code around the `require_tree` from [derailed_benchmarks](https://github.com/schneems/derailed_benchmarks)


QA
--
Checkout this branch and try running this with and without the script.  Test running it with the `time` command wrapping it and see if the sys/user times match up.  Try some of the other options available as well (FYI, to use `stackprof`/`sys/proctable`, they have to be included separately in your bundle for them to work).
